lib/srv/db/sqlserver/protocol/login7.go

```diff
@@ -143,3 +140,44 @@ func ReadLogin7Packet(r io.Reader) (*Login7Packet, error) {
 		database: database,
 	}, nil
 }
+
+// errInvalidPackage is returned when Login7 package contains invalid data.
+var errInvalidPackage = trace.Errorf("invalid login7 packet")
+
+// readUsername reads username from login7 package.
+func readUsername(pkt *Packet, header Login7Header) (string, error) {
+	if len(pkt.Data) < int(header.IbUserName)+int(header.CchUserName)*2 {
+		return "", errInvalidPackage
+	}
+
+	if len(pkt.Data) <= int(header.IbUserName) {
+		return "", errInvalidPackage
+	}
+
+	// Decode username and database from the packet. Offset/length are counted
+	// from the beginning of entire packet data (excluding header).
+	username, err := mssql.ParseUCS2String(
+		pkt.Data[header.IbUserName : header.IbUserName+header.CchUserName*2])
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return username, nil
+}
+
+// readDatabase reads database name from login7 package.
+func readDatabase(pkt *Packet, header Login7Header) (string, error) {
+	if len(pkt.Data) < int(header.IbDatabase)+int(header.CchDatabase)*2 {
+		return "", errInvalidPackage
+	}
+
+	if len(pkt.Data) < int(header.IbDatabase) {
+		return "", errInvalidPackage
+	}
```

Comment:
The same as above. Removed, but here is why `<` is required in this case:
```
% go test -count=1 ./...
--- FAIL: TestReadLogin7 (0.00s)
    protocol_test.go:50: 
                Error Trace:    protocol_test.go:50
                Error:          Received unexpected error:
                                invalid login7 packet
                Test:           TestReadLogin7
FAIL
FAIL    github.com/gravitational/teleport/lib/srv/db/sqlserver/protocol 0.496s
?       github.com/gravitational/teleport/lib/srv/db/sqlserver/protocol/fixtures        [no test files]
FAIL
```
Otherwise the `TestReadLogin7` fails as it tries to read the last element from the slice (it won't panic. In this case last element is bounded by last element here: `pkt.Data[136:136]` with len 136).